<p>
	Arrow functions are an ES6 feature.
	They are not part of the ES5 specification, and are unsupported by many browsers.
	An anonymous function should be used instead.
</p>
<h2>Noncompliant Code Example</h2>
<pre>
	[].forEach(item => console.log);
</pre>
<h2>Compliant Solution</h2>
<pre>
	[].forEach(function(item){
		console.log(item);
	});
</pre>
