<p>
	<code>const</code> and <code>let</code> variable declarations are an ES6 feature.
	They are not part of the ES5 specification, and are unsupported by many browsers.
	A normal <code>var</code> declaration should be used instead.
</p>
<h2>Noncompliant Code Example</h2>
<pre>
	const CONST = 3.14;

	function foo() {
		let a = 5;
	}
</pre>
<h2>Compliant Solution</h2>
<pre>
	var CONST = 3.14;

	function foo() {
		var b = 5;
	}
</pre>