<p>
	Iterators are an ES6 feature.
	They are not part of the ES5 specification, and are unsupported by many browsers.
	A <code>for-in</code> loop should be used to iterate over objects, while a <code>for/while</code> loop
	or <code>Array.forEach</code> should be used to iterate over arrays.
</p>
<h2>Noncompliant Code Example</h2>
<pre>
	let a = [];

	for (let item of a) {

	}
</pre>
<h2>Compliant Solution</h2>
<pre>
	var a = [];

	a.forEach(function(item) {

	});
</pre>