<p>
	Binary and octal numeric literals are an ES6 feature.
	They are not part of the ES5 specification, and are unsupported by many browsers.
	A decimal or hexadecimal numeric literal should be used instead.
</p>
<h2>Noncompliant Code Example</h2>
<pre>
	let
		a = 0b0110,
		b = 0o01;
</pre>
<h2>Compliant Solution</h2>
<pre>
	var
		a = 0x5,
		b = 1;
</pre>