<p>
	Object property shorthands and variable keys are an ES6 feature.
	They are not part of the ES5 specification, and are unsupported by many browsers.
	Pair properties with strong literal or identifier keys should be used instead.
</p>
<h2>Noncompliant Code Example</h2>
<pre>
	let o = {
		x, y
		z() {},
	};
</pre>
<h2>Compliant Solution</h2>
<pre>
	var o = {
		x: x,
		y: y,
		z: function() {}
	};
</pre>